### ISSUE TITLE
26.3; pyls not present on the path for  Win10 Emacs 26.3 and cygwin pyls

### ISSUE THREAD

**Issue by sivaramn** [emacs-lsp/lsp-mode#1976]
````
In GNU Emacs 26.3 (build 1, x86_64-w64-mingw32)
 of 2019-08-29 built on CIRROCUMULUS
Repository revision: 96dd0196c28bc36779584e47fffcca433c9309cd
Windowing system distributor 'Microsoft Corp.', version 10.0.18363
Recent messages:
LSP :: The following servers support current file but do not have automatic installation configuration: pyls
You may find the installation instructions at https://emacs-lsp.github.io/lsp-mode/page/languages.
(If you have already installed the server check *lsp-log*).
You can run the command ‘lsp’ with s-l s s
Configured using:
 'configure --without-dbus --host=x86_64-w64-mingw32
 --without-compress-install 'CFLAGS=-O2 -static -g3''

Configured features:
XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB
TOOLKIT_SCROLL_BARS THREADS LCMS2

Important settings:
  value of $LANG: ENN
  locale-coding-system: utf-8
````
I have `python` installed in Cygwin64 and I'm able to execute python scripts when opened in Win64 emacs 26.3 through an inferior shell process.  After installing `lsp-mode` and configuring `lsp-pyls-server-command` to any of `pyls` , `/usr/local/bin/pyls` , `/usr/bin/pyls`, the `*lsp-log*` is the same `Command "pyls" is not present on the path.`  I of course have `pyls` installed in all the above paths in cygwin.  And the cygwin version of emacs works correctly.  I can see the  `lsp-log` having messages like 
````
Command "pyls" is present on the path.
Command "pyls" is present on the path.
Found the following clients for /home/user/a.py: (server-id pyls, priority -1)
The following clients were selected based on priority: (server-id pyls, priority -1)
````
and python hinting of various sorts works.

This does not work in the Win10 version.  I tried variations of  `(setq lsp-pyls-server-command "C:/cygwin64/bin/pyls")` and that didn't work either.

Anything else I should check out?

**Comment by yyoncho** [comment]
You should make sure that (executable-find "path-to-pyls") is working.
What I would try is to go with dired in pyls dir, and do "C-0 w" which
will copy the path to pyls and set it to lsp-pyls-server-command.

**Comment by sivaramn** [comment]
> (executable-find "path-to-pyls")

Weird. Running it against a few binaries found in cygwin `bin` and `/usr/local/bin` I get the paths for some but I don't get it for `pyls`
```
"c:/cygwin64/usr/local/bin/tmux.exe"
"c:/cygwin64/bin/zip.exe"
"c:/cygwin64/bin/zsh.exe"
nil
````
The last one happened for `pyls`, `pylint` and `pyflake` for `executable-find`

and 

````
exec-path is a variable defined in ‘C source code’.
Its value is
("/usr/local/bin" "c:/cygwin64/bin" "c:/windows/system32" "C:/windows"....
````

Now, even more confused

**Comment by yyoncho** [comment]
IDK, maybe executable flags are not set?

**Comment by sivaramn** [comment]
> 
> 
> IDK, maybe executable flags are not set?

All are 755 mode.  I checked that too.

**Comment by kiennq** [comment]
If you're using Windows 10, you probably better using WSL instead of cygwin.

Anyway, can you check if you can run `pyls` via `cmd.exe`?
If `pyls` can only run inside `cygwin` then you have to use `cygwin` Emacs instead of `msys` one.

**Comment by yyoncho** [comment]
Any update here?

**Comment by sivaramn** [comment]
> 
> 
> If you're using Windows 10, you probably better using WSL instead of cygwin.
> 
> Anyway, can you check if you can run `pyls` via `cmd.exe`?
> If `pyls` can only run inside `cygwin` then you have to use `cygwin` Emacs instead of `msys` one.

`pyls` does **not** run within `cmd.exe`.  `pyls` does work for me under cygwinized Emacs.  The issue seems to be weird, that some cygwin binaries are executed by Win10 Emacs64 but some aren't.  Like I can do  `shell-command` from win10 and call `pyls` and I do get `(Shell command succeeded with no output)`.   So I can do a shell escape to call it from within emacs **but not** elispy way in `lsp-mode`

**Comment by sivaramn** [comment]
> 
> 
> Any update here?

Sorry for the late reply, answered to prior poster

**Comment by kiennq** [comment]
Well, Cygwin is kind of a POSFIX layer on Windows, and Cygwin binaries normally cannot run directly from Windows without using compatible shell.
So if you use `shell-command` (and set correct shell), Windows Emacs can execute Cygwin one.
`lsp-mode` use `make-process` which will not call process via any shell, so it will not work for you.

How about customize `lsp-clients-python-command` to include Cygwin shell and pass additional pyls parameters there.
It can work if the `stdio` communication with that shell and `lsp-mode` works

**Comment by sivaramn** [comment]
> How about customize `lsp-clients-python-command` to include Cygwin shell and pass additional pyls parameters there.
> It can work if the `stdio` communication with that shell and `lsp-mode` works

I already tried this configuring `lsp-pyls-server-command` to any of `pyls` , `/usr/local/bin/pyls` , `/usr/bin/pyls` and that hasn't worked. Are you suggesting something like `C:\cygwin64\bin\mintty.exe -i` with some parameters?  What would it be like? I choose `mintty` as a random example.

**Comment by kiennq** [comment]
> Are you suggesting something like `C:\cygwin64\bin\mintty.exe -i` with some parameters? What would it be like? I choose `mintty` as a random example.

Yes, but a shell (like `bash.exe`), not a terminal like `mintty`.

**Comment by kiennq** [comment]
Although if you're on Windows 10, something like [WSL](https://docs.microsoft.com/en-us/windows/wsl/wsl2-index) will fare much more better than Cygwin

**Comment by sivaramn** [comment]
> 
> 
> > Are you suggesting something like `C:\cygwin64\bin\mintty.exe -i` with some parameters? What would it be like? I choose `mintty` as a random example.
> 
> Yes, but a shell (like `bash.exe`), not a terminal like `mintty`.

yes, I get that.  What should I be plugging in then as `Command "C:\cygwin64\bin\pyls" is not present on the path` is what I get if I configure (setq lsp-pyls-server-command "C:\\cygwin64\\bin\\pyls").
